File: js_version/src/NeuralNetwork.js

'use strict'


var Neuron = require("./Neuron.js");


//
//
// NET

function NeuralNetwork (arr_topology) { // ctor/dtor

	// attr -> error

    this._m_error = 0.0;
    this._m_recentAvgError = 0.0;

	// attr

	this._m_arr_layers = []; // m_layers[layerNum][neuronNum]

	if (arr_topology.length == 0)
		throw new Error('(arr_topology.length == 0)');

    for (var i = 0; i < arr_topology.length; ++i)
    {
        var num_neuron = arr_topology[i];

		if (num_neuron <= 0) // no empty layer
			throw new Error('(num_neuron <= 0)');

        var arr_new_layer = [];

        //
        //

        var is_last_layer = (i == (arr_topology.length - 1));

        // 0 output if on the last layer
        var numOutputs = ((is_last_layer) ? (0) : (arr_topology[i + 1]));

        // We have a new layer, now fill it with neurons, and
        // add a bias neuron in each layer.
        for (var j = 0; j < (num_neuron + 1); ++j) // add a bias neuron
            arr_new_layer.push( new Neuron(numOutputs, j) );

        // Force the bias node's output to 1.0 (it was the last neuron pushed in this layer):
        var bias_neuron = arr_new_layer[ arr_new_layer.length - 1 ];
        bias_neuron.setOutputVal(1.0);
        // bias_neuron._m_outputVal = 1.0;

        //
        //

        this._m_arr_layers.push(arr_new_layer);
    }
};

// static attr -> error

NeuralNetwork.prototype._k_recentAvgSmoothingFactor = 100.0; // Number of training samples to average over

// public method(s)

NeuralNetwork.prototype.feedForward = function(arr_inputVals)
{
	if (arr_inputVals.length != (this._m_arr_layers[0].length - 1)) // exclude bias neuron
		throw new Error('(arr_inputVals.length != (m_arr_layers[0].length - 1))');

    // Assign (latch) the input values into the input neurons
    for (var i = 0; i < arr_inputVals.length; ++i)
        this._m_arr_layers[0][i].setOutputVal(arr_inputVals[i]);
        // this._m_arr_layers[0][i]._m_outputVal = arr_inputVals[i];


    // console.log(this._m_arr_layers[0]);

    // forward propagate
    for (var i = 1; i < this._m_arr_layers.length; ++i) // exclude input layer
    {
        var arr_prevLayer = this._m_arr_layers[i - 1];
        var arr_currLayer = this._m_arr_layers[i];

        var num_neuron = (arr_currLayer.length - 1); // exclude bias neuron
        for (var n = 0; n < num_neuron; ++n)
            arr_currLayer[n].feedForward(arr_prevLayer);
    }
}

NeuralNetwork.prototype.backProp = function(arr_targetVals)
{
    //
    // error

    // Calculate overall net error (RMS of output neuron errors)

    var arr_outputLayer = this._m_arr_layers[this._m_arr_layers.length - 1];
    this._m_error = 0.0;

    for (var n = 0; n < arr_outputLayer.length - 1; ++n)
    {
        var delta = arr_targetVals[n] - arr_outputLayer[n].getOutputVal();
        this._m_error += delta * delta;
    }
    this._m_error /= (arr_outputLayer.length - 1); // get average error squared
    this._m_error = Math.sqrt(this._m_error); // RMS

    // Implement a recent average measurement

    this._m_recentAvgError =
            (this._m_recentAvgError * this._k_recentAvgSmoothingFactor + this._m_error)
            / (this._k_recentAvgSmoothingFactor + 1.0);

    // error
    //


    //
    // Gradients

    // Calculate output layer gradients

    for (var n = 0; n < (arr_outputLayer.length - 1); ++n)
        arr_outputLayer[n].calcOutputGradients(arr_targetVals[n]);

    // Calculate hidden layer gradients

    for (var i = (this._m_arr_layers.length - 2); i > 0; --i)
    {
        var arr_hiddenLayer = this._m_arr_layers[i];
        var arr_nextLayer = this._m_arr_layers[i + 1];

        for (var n = 0; n < arr_hiddenLayer.length; ++n)
            arr_hiddenLayer[n].calcHiddenGradients(arr_nextLayer);
    }

    // Gradients
    //


    // For all layers from outputs to first hidden layer,
    // update connection weights

    for (var i = (this._m_arr_layers.length - 1); i > 0; --i)
    {
        var arr_currLayer = this._m_arr_layers[i];
        var arr_prevLayer = this._m_arr_layers[i - 1];

        for (var n = 0; n < (arr_currLayer.length - 1); ++n) // exclude bias
            arr_currLayer[n].updateInputWeights(arr_prevLayer);
    }
}

NeuralNetwork.prototype.getResults = function(arr_resultVals)
{
    arr_resultVals.length = 0;

    var arr_outputLayer = this._m_arr_layers[ this._m_arr_layers.length - 1 ];

    // exclude last neuron (bias neuron)
    var total_neuron = (arr_outputLayer.length - 1);

    for (var n = 0; n < total_neuron; ++n)
        arr_resultVals.push(arr_outputLayer[n].getOutputVal());
}

NeuralNetwork.prototype.getError = function() { return this._m_error; }
NeuralNetwork.prototype.getRecentAverageError = function() { return this._m_recentAvgError; }


NeuralNetwork.prototype.getWeights = function()
{
    var out_arr_weight = [];
    for (var i = 0; i < this._m_arr_layers.length - 1; ++i)
    {
        var curr_layer = this._m_arr_layers[i];

        var num_neurons = (curr_layer.length - 1); // exluding bias neuron

        for (var j = 0; j < num_neurons; ++j)
        {
            var curr_neuron = curr_layer[j];

            var arr_weight = curr_neuron.getOutputWeight();

            for (var k = 0; k < arr_weight.length; ++k)
                out_arr_weight.push( arr_weight[k].weight );
        }
    }

    return out_arr_weight;
}

// NET
//
//


module.exports = NeuralNetwork;

File: projects/training-logic/src/machine-learning/NeuralNetwork.cpp

#include "NeuralNetwork.hpp"

#include <cassert>


double NeuralNetwork::k_recentAvgSmoothingFactor = 100.0; // Number of training samples to average over

NeuralNetwork::NeuralNetwork(const std::vector<uint32_t>& arr_topology)
    :   m_error(0.0),
        m_recentAvgError(0.0)
{
    assert( !arr_topology.empty() ); // no empty topology

    RandomNumberGenerator rng;
    rng.ensureRandomSeed();

    for (uint32_t ii = 0; ii < arr_topology.size(); ++ii)
    {
        uint32_t totalNeurons = arr_topology[ii];

        assert( totalNeurons > 0 ); // no empty layer

        totalNeurons += 1; // add a bias neuron

        m_arr_layers.push_back(t_Layer());

        t_Layer& arr_new_layer = m_arr_layers.back();
        arr_new_layer.reserve(totalNeurons); // pre-allocate

        const bool is_last_layer = ((ii + 1) == arr_topology.size());

        // 0 output if on the last layer
        const uint32_t numOutputs = ((is_last_layer) ? (0) : (arr_topology[ii + 1]));

        // We have a new layer, now fill it with neurons (+ the extra bias neuron)
        for (uint32_t jj = 0; jj < totalNeurons; ++jj)
        {
            arr_new_layer.emplace_back(numOutputs, jj, rng);
        }

        // Force the bias node's output to 1.0
        // -> it was the last neuron pushed in this layer
        Neuron& bias_neuron = arr_new_layer.back();
        bias_neuron.setOutputVal(1.0);
    }
}

void NeuralNetwork::feedForward(const t_vals &inputVals)
{
    assert( inputVals.size() == (m_arr_layers[0].size() - 1) ); // exclude bias neuron

    // Assign (latch) the input values into the input neurons
    for (uint32_t ii = 0; ii < inputVals.size(); ++ii)
    {
        m_arr_layers[0][ii].setOutputVal(inputVals[ii]);
    }

    // forward propagate
    // -> start at 1 -> exclude input layer
    for (uint32_t ii = 1; ii < m_arr_layers.size(); ++ii)
    {
        t_Layer& prevLayer = m_arr_layers[ii - 1];
        t_Layer& currLayer = m_arr_layers[ii];

        const uint32_t num_neuron = uint32_t(currLayer.size()) - 1; // exclude bias neuron
        for (uint32_t jj = 0; jj < num_neuron; ++jj)
        {
            currLayer[jj].feedForward(prevLayer);
        }
    }
}

void NeuralNetwork::backProp(const t_vals &arr_targetVals)
{
    //
    // error

    // Calculate overall net error (RMS of output neuron errors)

    t_Layer &outputLayer = m_arr_layers.back();
    m_error = 0.0;

    // exclude bias neuron
    const uint32_t numOutputs = uint32_t(outputLayer.size()) - 1;

    for (uint32_t ii = 0; ii < numOutputs; ++ii)
    {
        const double delta = arr_targetVals[ii] - outputLayer[ii].getOutputVal();
        m_error += delta * delta;
    }
    m_error /= (outputLayer.size() - 1); // get average error squared
    m_error = std::sqrt(m_error); // RMS

    // Implement a recent average measurement

    m_recentAvgError =
            (m_recentAvgError * k_recentAvgSmoothingFactor + m_error)
            / (k_recentAvgSmoothingFactor + 1.0);

    // error
    //


    //
    // Gradients

    // Calculate output layer gradients
    for (uint32_t ii = 0; ii < numOutputs; ++ii)
    {
        outputLayer[ii].calcOutputGradients(arr_targetVals[ii]);
    }

    // Calculate hidden layer gradients

    // apply -2 to remove the input and output layer
    // possibly 0
    const uint32_t numHidden = uint32_t(m_arr_layers.size()) - 2;

    for (uint32_t ii = numHidden; ii > 0; --ii)
    {
        t_Layer &arr_currLayer = m_arr_layers[ii];
        const t_Layer &arr_nextLayer = m_arr_layers[ii + 1];

        for (auto& currNeuron : arr_currLayer)
        {
            currNeuron.calcHiddenGradients(arr_nextLayer);
        }
    }

    // Gradients
    //


    // For all layers from outputs to first hidden layer,
    // update connection weights

    const uint32_t numInputAndHidden = numHidden + 1;

    for (uint32_t ii = numInputAndHidden; ii > 0; --ii)
    {
        t_Layer &arr_prevLayer = m_arr_layers[ii - 1];
        const t_Layer &arr_currLayer = m_arr_layers[ii];

        // exclude last neuron (bias neuron)
        const uint32_t sizeLayer = uint32_t(arr_currLayer.size()) - 1;

        for (uint32_t jj = 0; jj < sizeLayer; ++jj)
        {
            arr_currLayer[jj].updateInputWeights(arr_prevLayer);
        }
    }
}

void NeuralNetwork::getResults(t_vals &arr_resultVals) const
{
    arr_resultVals.clear();

    const t_Layer& outputLayer = m_arr_layers.back();

    // exclude last neuron (bias neuron)
    const uint32_t total_neuron = uint32_t(outputLayer.size()) - 1;

    arr_resultVals.reserve(total_neuron); // pre-allocate

    for (uint32_t ii = 0; ii < total_neuron; ++ii)
    {
        arr_resultVals.push_back(outputLayer[ii].getOutputVal());
    }
}
